### PR TITLE
Added option for generating 'struct' for types defined with 'struct' keyword wherever possible.

### PR DESCRIPTION
**Pull request by genuinelucifer**
Added everything and as faar as I can see it is working fine.
Please review @ddobrev @tritao

### CODE REVIEW

**Comment by genuinelucifer** [comment]
Had to change two tests from 'struct' to 'class' because they were generating erroneous code when marshalled with 'struct' keyword.

File: src/AST/PropertyExtensions.cs [hunk 1]
```diff
@@ -14,7 +14,7 @@ public static bool IsInRefTypeAndBackedByValueClassField(this Property p)
             type = type ?? p.Field.Type;
 
             Class decl;
-            return type.TryGetClass(out decl) && decl.IsValueType;
+            return type.TryGetClass(out decl) && (decl.IsValueType || decl.IsStruct);
```
Comment (ddobrev) [inline src/AST/PropertyExtensions.cs]: It seems to me that you use this || expression any time IsValueType is used. In that case why don't you move || decl.IsStruct to the getter of IsValueType analogous to what you did for IsRefType? This way we can continue checking just IsValueType and we'll also have many less changes in this pull.

File: src/CppParser/AST.h [hunk 2]
```diff
@@ -716,6 +716,7 @@ class CS_API Class : public DeclarationContext
     bool HasNonTrivialCopyConstructor;
     bool HasNonTrivialDestructor;
     bool IsExternCContext;
+	bool DeclaredStruct;
```
Comment (ddobrev) [inline src/CppParser/AST.h]: This is aligned with a tab instead of with 4 spaces as per convention.

File: src/CppParser/Bindings/CLI/AST.cpp [hunk 3]
```diff
@@ -2835,6 +2835,11 @@ void CppSharp::Parser::AST::Class::IsExternCContext::set(bool value)
     ((::CppSharp::CppParser::AST::Class*)NativePtr)->IsExternCContext = value;
 }
 
+bool CppSharp::Parser::AST::Class::DeclaredStruct::get()
+{
+	return ((::CppSharp::CppParser::AST::Class*)NativePtr)->DeclaredStruct;
```
Comment (ddobrev) [inline src/CppParser/Bindings/CLI/AST.cpp]: Misaligned.

File: src/CppParser/Bindings/CLI/AST.h [hunk 4]
```diff
@@ -1840,6 +1840,11 @@ namespace CppSharp
                     void set(bool);
                 }
 
+				property bool DeclaredStruct
```
Comment (ddobrev) [inline src/CppParser/Bindings/CLI/AST.h]: Misaligned.

File: src/CppParser/Parser.cpp [hunk 5]
```diff
@@ -685,6 +685,7 @@ void Parser::WalkRecord(clang::RecordDecl* Record, Class* RC)
     RC->IsUnion = Record->isUnion();
     RC->IsDependent = Record->isDependentType();
     RC->IsExternCContext = Record->isExternCContext();
+	RC->DeclaredStruct = Record->isStruct();
```
Comment (ddobrev) [inline src/CppParser/Parser.cpp]: Misaligned.

File: src/Generator/Generators/CLI/CLIHeadersTemplate.cs [hunk 6]
```diff
@@ -283,6 +286,34 @@ public void GenerateClass(Class @class)
...
+
+             if (@class.IsValueType)
+                 return true;
+             if (@class.IsInterface)
+                 return false;
+             if (@class.IsStatic)
+                 return false;
+             if (@class.IsAbstract)
+                 return false;
+             if (@class.HasBase)
```
Comment (ddobrev) [inline src/Generator/Generators/CLI/CLIHeadersTemplate.cs]: Are you sure this is necessary? C# structs can inherit from other classes, they just cannot be inherited from.

File: src/Generator/Generators/CSharp/CSharpTextTemplate.cs [hunk 7]
```diff
@@ -353,6 +353,12 @@ public void GenerateInlineSummary(RawComment comment)
 
         public void GenerateClass(Class @class)
         {
+            if(@class.DeclaredStruct)
```
Comment (ddobrev) [inline src/Generator/Generators/CSharp/CSharpTextTemplate.cs]: Please add a space after the 'if'.

File: src/Generator/Generators/CSharp/CSharpTextTemplate.cs [hunk 8]
```diff
@@ -353,6 +353,12 @@ public void GenerateInlineSummary(RawComment comment)
...
+            {
+                WriteLine("//Declared Struct");
```
Comment (ddobrev) [inline src/Generator/Generators/CSharp/CSharpTextTemplate.cs]: Not that it's a problem but what does this help with?

File: src/Generator/Generators/CSharp/CSharpTextTemplate.cs [hunk 9]
```diff
@@ -434,6 +440,34 @@ private void GenerateClassMarshals(Class @class)
             NewLine();
         }
 
+         private bool CheckClassIsStructible(Class @class)
```
Comment (ddobrev) [inline src/Generator/Generators/CSharp/CSharpTextTemplate.cs]: You have written this method for the second time. Please move it to some common to both generators location, such as ClassExtensions.

File: tests/Basic/Basic.h [hunk 10]
```diff
@@ -486,8 +486,9 @@ DLL_API int Function()
 }
 
 // Tests properties
-struct DLL_API TestProperties
+class DLL_API TestProperties
```
Comment (ddobrev) [inline tests/Basic/Basic.h]: Please add a few words explaining why this type has to be a class. I just want to be sure there's no hidden bug revealed by this test.

File: tests/CSharpTemp/CSharpTemp.h [hunk 11]
```diff
@@ -175,8 +175,9 @@ class DLL_API P : Proprietor
 };
 
 // Tests destructors
-struct DLL_API TestDestructors
+class DLL_API TestDestructors
```
Comment (ddobrev) [inline tests/CSharpTemp/CSharpTemp.h]: Same as the previous test: please add a few words to explain this change.

**Comment by genuinelucifer** [comment]
Instead of changing the tests I thought it's better to change the allocation of 'IsStruct' property.

File: src/Generator/Generators/CSharp/CSharpTextTemplate.cs (hunk 8 above)
Comment (genuinelucifer) [inline src/Generator/Generators/CSharp/CSharpTextTemplate.cs]: Sorry for this one, I forgot to erase it. I used it just to test if 'DeclaredStruct' was properly initialised.

File: src/Generator/Generators/CSharp/CSharpTextTemplate.cs [hunk 12]
```diff
@@ -434,6 +436,38 @@ private void GenerateClassMarshals(Class @class)
...
+                 return true;
+             if (@class.IsInterface)
+                 return false;
+             if (@class.IsStatic)
+                 return false;
+             if (@class.IsAbstract)
+                 return false;
+             if (@class.HasBase)
+                 return false;
+             if (@class.Methods.Any(m => m.IsStatic))
```
Comment (ddobrev) [inline src/Generator/Generators/CSharp/CSharpTextTemplate.cs]: C# structs are allowed to have static members and operators so this fix is incorrect. If the only problem with those tests is their static members, there's some other bug that needs fixing.

File: src/Generator/Generators/CLI/CLIHeadersTemplate.cs (hunk 6 above)
Comment (genuinelucifer) [inline src/Generator/Generators/CLI/CLIHeadersTemplate.cs]: Sorry. I don't even remember now why I added that at first. Will remove it.

File: src/AST/PropertyExtensions.cs (hunk 1 above)
Comment (genuinelucifer) [inline src/AST/PropertyExtensions.cs]: I first thought IsValueType and IsStruct were mean to be in different contexts. But now yes they seem to be same. Will move it.

**Comment by genuinelucifer** [comment]
@ddobrev Yes, the build was failing due to obe of the static members in ValueType..
Will try to fix it...

File: src/Generator/Generators/ExtensionMethods.cs [hunk 13]
```diff
@@ -23,5 +24,35 @@ public static Interop.CallingConvention ToInteropCallConv(this CallingConvention
...
+                return false;
+            if (@class.Methods.Any(m => m.IsStatic))
+                return false;
+            if (@class.Methods.Any(m => m.IsOperator))
+                return false;
+
+            var allTrUnits = Driver.ASTContext.TranslationUnits;
+            foreach (var trUnit in allTrUnits)
+            {
+                foreach (var cls in trUnit.Classes)
```
Comment (genuinelucifer) [inline src/Generator/Generators/ExtensionMethods.cs]: Please verify once that it is correct way to loop through ALL the classes. I am not sure but have found no error so far.

File: src/AST/Class.cs [hunk 14]
```diff
@@ -102,6 +102,9 @@ public class Class : DeclarationContext
         // True if the class represents a static class.
         public bool IsStatic;
 
+        // True if declared with 'struct' keyword
+        public bool DeclaredStruct;
```
Comment (ddobrev) [inline src/AST/Class.cs]: You haven't renamed this to "IsStruct" as I asked.

File: src/Generator/Generators/ExtensionMethods.cs (hunk 13 above)
Comment (ddobrev) [inline src/Generator/Generators/ExtensionMethods.cs]: Looks fine to me.

File: src/AST/Class.cs (hunk 14 above)
Comment (genuinelucifer) [inline src/AST/Class.cs]: I have used 'DeclaredStruct' which is filled by AST just to check if while declaring this 'struct' keyword was used.
'IsStruct' is used to determine if the type follows all rules and should be marshalled to 'struct'...
.
Should I combine them both?

Comment (ddobrev) [inline src/AST/Class.cs]: A good question. I think it it all works, you'd better combine them. Also, now that you're at it, please make it a property (best if automatic) and not a field. I know the rest are field but it's a bad practice and we should eliminate it over time.

File: src/AST/Class.cs [hunk 15]
```diff
@@ -149,6 +154,12 @@ public bool HasBaseClass
             get { return BaseClass != null; }
         }
 
+        public bool IsStruct
+        {
+            get;
+            set;
```
Comment (genuinelucifer) [inline src/AST/Class.cs]: @ddobrev By automatic property, do you mean something like this?

Comment (ddobrev) [inline src/AST/Class.cs]: Exactly.

File: src/AST/Class.cs (hunk 14 above)
Comment (genuinelucifer) [inline src/AST/Class.cs]: Wouldn't it seem erroneous to see some code like:
```C#
@class.IsStruct = CheckClassIsStructible(@class, Driver);
```
while the function itself uses the same property as:
```C#
if(!@class.IsStruct)
    return false;
```

Comment (ddobrev) [inline src/AST/Class.cs]: I wouldn't say so. Basically, C# structs are limited compared to C++ structs so this code simply means that we apply these limitations on top if the type is a (C++) struct.

Comment (genuinelucifer) [inline src/AST/Class.cs]: Ohh. Fine.

**Comment by genuinelucifer** [comment]
Although i combined them but now the CLI generator is generating erroneous code in Basic.cpp 'Exception' and STL.cpp
I am looking for the cause.

**Comment by ddobrev** [comment]
In case you have problem with this, wouldn't setting of Class.Type to ClassType.ValueType instead of setting IsStruct to true, work better?

**Comment by genuinelucifer** [comment]
@ddobrev Thanks. Using that and a few other bits reduced the errors a lot. I am working on removing the left errors. For that we need some changes which I am wroking on in #506 .

**Comment by genuinelucifer** [comment]
Once #506 is merged, i will open a fresh pull with all the working changes.
But I had to change the TestDestructors test from 'struct' to 'class' because structs can't have destructors in C#.
Should I improve the check to include this or we could provide a new macro CS_REF_TYPE.. as CS_VALUE_TYPE marshals a C++ class to C# struct.
CS_REF_TYPE could be used to force marshaling a C++ struct to C# class...

File: src/Generator/Generators/CLI/CLIHeadersTemplate.cs [hunk 16]
```diff
@@ -137,6 +137,9 @@ public void GenerateForwardRefs()
             var typeReferences = typeReferenceCollector.TypeReferences;
             var @namespace = ConvertForwardReferencesToNamespaces(typeReferences);
 
+            foreach (var cls in @namespace.Classes)
+                cls.IsStruct = ExtensionMethods.CheckClassIsStructible(cls, Driver);
```
Comment (tritao) [inline src/Generator/Generators/CLI/CLIHeadersTemplate.cs]: We should not be modifying the AST in the text generators. This should be done in a pass.

File: src/Generator/Generators/CSharp/CSharpTextTemplate.cs [hunk 17]
```diff
@@ -353,6 +353,8 @@ public void GenerateInlineSummary(RawComment comment)
 
         public void GenerateClass(Class @class)
         {
+            @class.IsStruct = ExtensionMethods.CheckClassIsStructible(@class, Driver);
```
Comment (tritao) [inline src/Generator/Generators/CSharp/CSharpTextTemplate.cs]: We should not be modifying the AST in the text generators. This should be done in a pass.
 Add a line note

File: src/Generator/Generators/ExtensionMethods.cs [hunk 18]
```diff
@@ -23,5 +24,30 @@ public static Interop.CallingConvention ToInteropCallConv(this CallingConvention
...
+        public static bool CheckClassIsStructible(Class @class, Driver Driver)
+        {
+            if (!@class.IsStruct)
+                return false;
+            if (@class.IsInterface || @class.IsStatic || @class.IsAbstract)
+                return false;
+            if (@class.Methods.Any(m => m.IsOperator))
+                return false;
+
+            if (@class.IsValueType)
```
Comment (tritao) [inline src/Generator/Generators/ExtensionMethods.cs]: I think this check should be done first and have priority to the other checks.

File: src/Generator/Generators/ExtensionMethods.cs [hunk 19]
```diff
@@ -23,5 +24,30 @@ public static Interop.CallingConvention ToInteropCallConv(this CallingConvention
...
             return Interop.CallingConvention.Winapi;
         }
+
+        public static bool CheckClassIsStructible(Class @class, Driver Driver)
+        {
+            if (!@class.IsStruct)
+                return false;
+            if (@class.IsInterface || @class.IsStatic || @class.IsAbstract)
+                return false;
+            if (@class.Methods.Any(m => m.IsOperator))
```
Comment (tritao) [inline src/Generator/Generators/ExtensionMethods.cs]: Why can't it be a struct if there are any operators? As far as I know C# structs also support operators.

**Comment by tritao** [comment]
Overall I don't think this approach makes sense, we can't just assume something is a struct just by looking if it was originally declared as a struct in C++. We should be instead checking its structure (operators, inheritance chain, members) and if it's a POD type.

**Comment by ddobrev** [comment]
What doesn't make sense? He does make all of the mentioned checks.

File: src/Generator/Generators/ExtensionMethods.cs (hunk 19 above)
Comment (ddobrev) [inline src/Generator/Generators/ExtensionMethods.cs]: Right, I've told already told him about it, not sure why he hasn't removed it yet.

File: src/Generator/Generators/CLI/CLIHeadersTemplate.cs (hunk 16 above)
Comment (genuinelucifer) [inline src/Generator/Generators/CLI/CLIHeadersTemplate.cs]: Should I add a new pass? I can't find any pass that would be fit to accommodate this.
Also, I can add a new file for the pass but do I actually get it working? Please tell me the file where it is defined as to which passes will be applicable and which not for a certain test project.

File: src/Generator/Generators/ExtensionMethods.cs (hunk 18 above)
Comment (genuinelucifer) [inline src/Generator/Generators/ExtensionMethods.cs]: Yes, I actually have changed many things from this pull but in a separate branch. I will do that here as well.

File: src/Generator/Generators/CLI/CLIHeadersTemplate.cs (hunk 16 above)
Comment (ddobrev) [inline src/Generator/Generators/CLI/CLIHeadersTemplate.cs]: I think you'd best move the whole logic about checking whether a type can be represented as a struct, to a new pass.

File: src/Generator/Generators/ExtensionMethods.cs (hunk 19 above)
Comment (genuinelucifer) [inline src/Generator/Generators/ExtensionMethods.cs]: Sorry, I was experimenting with different options and changes in a separate branch. I will apply the changes here as well.

**Comment by genuinelucifer** [comment]
@tritao I have pushed a couple of changes here. But, I will need the changes of #506 for these to work correctly. I will submit another PR with fresh changes.
.
Also, I too don't see why it is not making sense?
We have all the proper checks here before deciding. The 'struct' keyword is just one of those checks!

File: src/Generator/Generators/CLI/CLIHeadersTemplate.cs (hunk 16 above)
Comment (genuinelucifer) [inline src/Generator/Generators/CLI/CLIHeadersTemplate.cs]: Ok. Will do.
But how do I make sure that pass is applied to say Basic.h in the tests!?

File: tests/CSharpTemp/CSharpTemp.h [hunk 20]
```diff
@@ -175,8 +175,9 @@ class DLL_API P : Proprietor
 };
 
 // Tests destructors
-struct DLL_API TestDestructors
+class DLL_API TestDestructors                                  //Value Types can't have destructors
 {
+public:
```
Comment (genuinelucifer) [inline tests/CSharpTemp/CSharpTemp.h]: @ddobrev @tritao I know I should not have changed this. But, I tried a few things but could not find how to check if a 'destructor' or a 'zero argument Constructor' has been "explicitly defined" by the user! Please help.

File: tests/CSharpTemp/CSharpTemp.h [hunk 21]
```diff
@@ -175,8 +175,9 @@ class DLL_API P : Proprietor
 };
 
 // Tests destructors
-struct DLL_API TestDestructors
+class DLL_API TestDestructors                                  //Value Types can't have destructors
```
Comment (genuinelucifer) [inline tests/CSharpTemp/CSharpTemp.h]: @ddobrev @tritao I know I should not have changed this. But, I tried a few things but could not find how to check if a 'destructor' or a 'zero argument Constructor' has been "explicitly defined" by the user! Please help.

File: src/Generator/Generators/CLI/CLIHeadersTemplate.cs (hunk 16 above)
Comment (ddobrev) [inline src/Generator/Generators/CLI/CLIHeadersTemplate.cs]: You just add the pass to the list of passes in Driver.SetupPasses.

Comment (genuinelucifer) [inline src/Generator/Generators/CLI/CLIHeadersTemplate.cs]: Thanks. Will try to implement it.

File: tests/CSharpTemp/CSharpTemp.h (hunk 21 above)
Comment (tritao) [inline tests/CSharpTemp/CSharpTemp.h]: I think you can use clang's `Decl::isImplicit()` to check if the declaration was explicitly declared by the user.

**Comment by tritao** [comment]
How a type was written in native code (whether with `class` or `struct` keyword) should not make any difference to how its classified (`ref` vs `value` type). We should solely look at its structure and usage to make that determination.

File: tests/CSharpTemp/CSharpTemp.h (hunk 21 above)
Comment (genuinelucifer) [inline tests/CSharpTemp/CSharpTemp.h]: Thanks... will try that.

**Comment by genuinelucifer** [comment]
@tritao I have added a new pass to work for that and also I have removed the checking of 'struct' keyword in a separate branch. Please merge #506 first if you have no objections on the way that is done.
Then I will open the pull for this because I really need the changes on that pull.
Till then I am closing it since it is obsolete I looked at the changes and they are confusing here.

**Comment by ddobrev** [comment]
@tritao I think it should make a difference. The way we do it now does not check if a type's been written as a struct, it's actully the other way around. It's checked if it's a class - in which case it's left alone. If it's written as a struct, only then we apply the checks to see if it can be a struct.
@genuinelucifer when you get back to this issue, please add an additional check - if the type has any virtual functions.

**Pull request closed without merging**